### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by beroset**
This includes a previous pull request https://github.com/AsteroidOS/asteroidsyncservice/pull/47 but can be rebased if needed.  It adds extensive developer documentation and updates the README.md file.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,27 +5,73 @@ asteroidsyncserviced is a daemon relying on libasteroid to synchronize data from
...
+
+```
+sudo cmake --build build -j -t install
+```
+
+The `-t` option tells CMake the build target that we want.  In this case the target is named `install`. 
+
+### Building developer documentation (optional)
+
+To build the documentation for the project in either HTML or PDF format requires an additional option for the "pPrepare the build" step listed above.  The setting is `WITH_DOC` which is set to `OFF` by default, but can be turned on:
```
Comment (MagneFire) [inline README.md]: tTypo here

Comment (MagneFire) [inline README.md]: I'm not a native speaker. But this sentence is worded a bit odd (between specifying the formats and the additional option)?

Maybe something like: This project provides developer documentation in the HTML or PDF format. This is disabled by default. One can enable the generation of the documentation by adding an option during the "Prepare the build" steps listed above. To turn on the documentation generation use:

File: README.md [hunk 2]
```diff
@@ -5,27 +5,73 @@ asteroidsyncserviced is a daemon relying on libasteroid to synchronize data from
...
-Create build dir
+After the build is prepared, one can build the documentation from the top level directory:
 
-  - `mkdir build && cd build`
+```
+cmake --build build -j -t doc
+``` 
 
-Run qmake:
+This builds the software documentation in html format, the software is built in the `build/doc/html` subdirectory.  So to view the documents using Firefox, for example, one can use the command `firefox build/doc/html/index.html` which will load the main page.  Of course any browser can be used, and not just Firefox.
```
Comment (MagneFire) [inline README.md]: Inconsistent use of HTML or html.

Comment (MagneFire) [inline README.md]: **So** seems unnecessary here?

Comment (MagneFire) [inline README.md]: Do I see a double space before _Of course..._?

File: doc/01_intro.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# Introduction {#mainpage} #
+The purpose of a [synchronization client](@ref synchronization_client) for an AsteroidOS watch is to provide a means by which data and commands from the machine, generally, a computer or mobile phone running Linux on which the client runs, to the watch.  Although either a computer or mobile phone are commonly used for this, the rest of this document will refer to "the computer" to mean the computer on which the synchronization client runs.
```
Comment (MagneFire) [inline doc/01_intro.md]: In my opinion, this sounds a bit vague, even though you try to clarify things. The use of _the machine_ makes it more strange?
At first read gives the impression that the _synchronization client_ runs on the watch, which isn't the case.

I would change this to start from an even simpler level.

Something like this maybe?
AsteroidOS synchronizes notifications, weather data, music metadata, etc to the watch using a synchronization client running on a Linux computer or a mobile phone.

File: doc/01_intro.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+It's important to understand that this current work only provides *half* of the required software, and that a separate synchronization client then uses D-Bus to actually perform the synchronization.
```
Comment (MagneFire) [inline doc/01_intro.md]: This _separate synchronization client_ runs of the watch right?

File: doc/01_intro.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+For all of these, the communications between the watch and the computer uses [Bluetooth](https://www.bluetooth.com/learn-about-bluetooth/tech-overview/).  Later pages describe the details of this and how it works. 
+
+## An example
+
+An example may be useful to clarify.  In this example, we trace a single request to turn the volume up via the watch's music app.  From the user's point of view, here's a very simplified sequence diagram.  This assumes that the sofware is already running on the computer, that the watch is [paired](@ref paired), and that the watch is [connected](@ref connected).  The music player shown here isn't any particular software, but a class of software that can play media.  Real examples of such software include [vlc](https://www.videolan.org/) and [Rhythmbox](https://wiki.gnome.org/Apps/Rhythmbox), but as we will see later, as long as the software implements a particular kind of interface, it can be used.
+
+\startuml
```
Comment (MagneFire) [inline doc/01_intro.md:16]: Very cool! Didn't know this was even possible 🚀

File: doc/01_intro.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
...-based software packages.  In this instance, BlueZ handles many of the underlying complexity of Bluetooth for us.  The AsteroidOS uses [Bluetooth Low Energy](@ref ble) or BLE, to communicate.  BLE uses the General Attribute Profile (GATT) which collects individual data items called [characteristics](@ref characteristic) into [services](@ref service).  The `asteroidsyncserviced` software subscribes to notifications of changes for particular characteristics of interest.  When a characteristic changes, the software is notified.  In this particular example, that change is received by the MediaService::onCharacteristicChanged function. The notification also contains a passed value which is the desired volume level percentage expressed as an integer from 0 to 100.  This results in a signal, MediaService::volume being emitted, which is connected to the Platform::onMediaVolumeChange function.  That function then uses D-Bus once again to send a request to a media player to set its volume.  Unlike the MediaService class, however, the interface for the media player expects to receive the volume setting as a `double` in the range 0.0 to 1.0 so we divide by 100 to scale the value appropriately.
```
Comment (MagneFire) [inline doc/01_intro.md]: First sentence has mixed use of singular and plural construction.
(_a D-Bus messages_)

Comment (MagneFire) [inline doc/01_intro.md]: Use of **The** seems not needed here?
(_The AsteroidOS uses..._)

Comment (MagneFire) [inline doc/01_intro.md]: _asteroidsyncserviced_ first occurrence of the service with a postfixed with the letter **d**.

File: doc/99_glossary.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
+# Glossary {#glossary} #
+This glossary gives brief descriptions of some of the key terms used in this software.
+
+### Bluetooth Low Energy (BLE) {#ble} ###
+Bluetooth Low Energy is a specialization of the Bluetooth specification that is designed to be used with devices that need to conserve power.  Since AsteroidOS is designed to run on watches that run on batteries, using BLE is a natural fit.  See https://en.wikipedia.org/wiki/Bluetooth_Low_Energy for more details.
+
+### characteristic {#characteristic} ###
+a single data value in BLE that is transferred between devices.  An example is the current volume setting of media player software.
```
Comment (MagneFire) [inline doc/99_glossary.md]: This and the next sections start with a lower case. Is this on purpose?

File: doc/99_glossary.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+### connected {#connected} ###
+devices which are paired and both ready to send or receive data are connected.  Note that it is possible for devices to be paired but not connected, but they cannot be connected and unpaired.
```
Comment (MagneFire) [inline doc/99_glossary.md]: Nit pick. One cannot send and receive at the same time (and/or might be better)?

File: doc/99_glossary.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+### paired {#paired} ###
+two Bluetooth devices which have already completed pairing are said to be paired.
+
+### pairing {#pairing} ###
+the process by which a computer and a watch are "introduced" to each other so that they can begin to communicate via Bluetooth
```
Comment (MagneFire) [inline doc/99_glossary.md]: Misses a **.**

File: doc/99_glossary.md [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+devices which are paired and both ready to send or receive data are connected.  Note that it is possible for devices to be paired but not connected, but they cannot be connected and unpaired.
+
+### paired {#paired} ###
+two Bluetooth devices which have already completed pairing are said to be paired.
+
+### pairing {#pairing} ###
+the process by which a computer and a watch are "introduced" to each other so that they can begin to communicate via Bluetooth 
+
+### service {#service} ###
+In the context of BLE, a service can be though of as a collection of related characteristics.
```
Comment (MagneFire) [inline doc/99_glossary.md]: _though of_ or _thought of_?

**Review by MagneFire — COMMENTED** [review]
Thank you so much for this!
This is a very welcome addition and it helps current and future developers to get started with working on the synchronization client.

Left some comments. I should probably add that some may or may not make sense as I'm not a native speaker. Just ignore my silly mistakes please 😉

File: README.md (hunk 1 above)
Comment (beroset) [reply README.md]: tThanks! Fixed.

Comment (beroset) [reply README.md]: You're right.  I took your suggestion and trimmed even further.

File: README.md (hunk 2 above)
Comment (beroset) [reply README.md]: Good catch.  Fixed.

Comment (beroset) [reply README.md]: On second thought, the whole paragraph about viewing with Firefox seemed more than needed, so I removed most of the paragraph.

File: doc/01_intro.md (hunk 3 above)
Comment (beroset) [reply doc/01_intro.md]: Really good point.  I tried a modification of your suggestion.  Please let me know what you think.

File: doc/01_intro.md (hunk 4 above)
Comment (beroset) [reply doc/01_intro.md]: No, it runs on the computer.  That is really a terrible couple of paragraphs which I have now rewritten.

File: doc/01_intro.md (hunk 5 above)
Comment (beroset) [reply doc/01_intro.md:16]: It is.  I have tested both!

File: doc/01_intro.md (hunk 6 above)
Comment (beroset) [reply doc/01_intro.md]: Good catch!  Fixed.

Comment (beroset) [reply doc/01_intro.md]: I changed it to *asteroidsyncservice* daemon for clarity.

File: doc/99_glossary.md (hunk 8 above)
Comment (beroset) [reply doc/99_glossary.md]: I think you're right. I've used your suggested wording.

File: doc/99_glossary.md (hunk 9 above)
Comment (beroset) [reply doc/99_glossary.md]: I've tried to make all of the entries more consistent.  Now they are all proper sentences.

File: doc/99_glossary.md (hunk 10 above)
Comment (beroset) [reply doc/99_glossary.md]: You're right.  Fixed.

File: doc/99_glossary.md (hunk 7 above)
Comment (beroset) [reply doc/99_glossary.md]: It was intentional, but I've changed my mind and now all of the definitions are complete sentences.

File: doc/01_intro.md (hunk 6 above)
Comment (beroset) [reply doc/01_intro.md]: Right.

File: README.md (hunk 2 above)
Comment (beroset) [reply README.md]: I've reduced the paragraph to a single sentence now.

**Comment by beroset** [comment]
Thanks for the thorough reading and review.  I've accepted and (I hope!) improved almost all of your suggestions and I think it's better now as a result.  All of the silly errors were mine, and not yours!

File: doc/01_intro.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
+# Introduction {#mainpage} #
+The purpose of a [synchronization client](@ref synchronization_client) is to synchronize data between a device (a computer, phone or tablet) and a watch running AsteroidOS. Although either a computer or mobile phone are commonly used for this, the rest of this document will refer to "the computer" to mean the computer on which the synchronization client runs.
+
+It's important to understand that this current doesn't implement an entire synchronization client but provides relatively portable code on which a sychronization client can be built.  Two such examples are Telescope which runs on Ubuntu Touch, and starship which runs on SailfishOS.
```
Comment (FlorentRevest) [inline doc/01_intro.md]: s/this current/this project/ ?

s/starship/Starship/

File: doc/01_intro.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+The kinds of things that are synchronized can include things like the time of day and weather data for the next few days, or commands such as taking a screenshot of the watch face from your phone or turning the volume of music playing on your computer up or down from your watch.
```
Comment (FlorentRevest) [inline doc/01_intro.md]: s/things like// (avoids the repetition of "things")

File: doc/01_intro.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
...r Linux-based software packages.  In this instance, BlueZ handles many of the underlying complexity of Bluetooth for us.  AsteroidOS uses [Bluetooth Low Energy](@ref ble) or BLE, to communicate.  BLE uses the General Attribute Profile (GATT) which collects individual data items called [characteristics](@ref characteristic) into [services](@ref service).  The `asteroidsyncservice` daemon subscribes to notifications of changes for particular characteristics of interest.  When a characteristic changes, the software is notified.  In this particular example, that change is received by the MediaService::onCharacteristicChanged function. The notification also contains a passed value which is the desired volume level percentage expressed as an integer from 0 to 100.  This results in a signal, MediaService::volume being emitted, which is connected to the Platform::onMediaVolumeChange function.  That function then uses D-Bus once again to send a request to a media player to set its volume.  Unlike the MediaService class, however, the interface for the media player expects to receive the volume setting as a `double` in the range 0.0 to 1.0 so we divide by 100 to scale the value appropriately.
```
Comment (FlorentRevest) [inline doc/01_intro.md]: s/a message manager that allows Linux-based software packages to talk to other Linux-based software packages/an Inter-Process-Communication system for Linux processes/
(IPC is a fairly standard term)

If you surround code snippets (like MediaService::onCharacteristicChanged) with backticks they should be nicely printed out `in monospace`

The API we use to communicate with media players is actually a freedesktop standard called MPRIS fyi

File: doc/03_dbus.md [hunk 14]
```diff
@@ -0,0 +1,18 @@
+# Understanding D-Bus {#dbus} #
+This is not a comprehensive tutorial on D-Bus, but an overview of generally what D-Bus is, how it works, and how it is used within `asteroidsyncservice`.  For a more thorough view, see https://www.freedesktop.org/wiki/IntroductionToDBus/
+
+D-Bus is a way for different pieces of software to communicate with each other in a standarized way.  This is how the `asteroidsyncservice` software can, for instance, turn the volume up or down on media player software without needing to know exactly which media player software is being used.\footnote{Although the StarfishOS supports D-Bus, it is not used for this purpose on that platform}
```
Comment (FlorentRevest) [inline doc/03_dbus.md]: Regarding the footnote: actually it is, it's just encapsulated by libwatchfishhttps://github.com/piggz/libwatchfish/blob/master/musiccontroller.cpp#L364

File: doc/02_architecture.md [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    Component.onDestruction: serviceController.stopService()
+}
+```
+
+In the SyncServicePlugin::registerTypes function, the plugin exports three types: Watch, Watches and ServiceControl which is exported to QML as `ServiceController` as shown above.  Once the `ServiceController` is started, the Watches collection of Watch objects may be used to interact with a selected watch.
+
+Consult the details of those classes to read about the signals and properties that are available to a QML application.
+
+## Platform support ##
+The QML plugin provides two different ways to automatically start the daemon.  One is via [D-Bus](@ref dbus), which is used by the SailfishOS platform and the Linux desktop platform, and the other is via `upstart` which is used by the Ubuntu Touch platform.
```
Comment (FlorentRevest) [inline doc/02_architecture.md:37]: Iiuc all these "@ref" are supposed to be accompanied by glossary entries ? Are these entries you forgot to document or do I misunderstand @ref ?

**Comment by FlorentRevest** [comment]
Also I think it would be appropriate to link to https://asteroidos.org/wiki/ble-profiles/ when talking about BLE characteristics etc...

And one big thing I'd expect from this documentation is a clear distinction between "asteroidsyncserviced" the daemon and "asteroidsyncservice" the QML plugin which makes it easy to talk to "asteroidsyncserviced", I didn't find it particularly obvious from a first read.

File: doc/01_intro.md (hunk 11 above)
Comment (moWerk) [reply doc/01_intro.md]: Or maybe a `currently` or `this project currently`?

Comment (beroset) [reply doc/01_intro.md]: I think I'll go with "this project."  Thanks!

File: doc/01_intro.md (hunk 12 above)
Comment (beroset) [reply doc/01_intro.md]: Excellent!  Much better that way.

File: doc/01_intro.md (hunk 13 above)
Comment (beroset) [reply doc/01_intro.md]: I like that edit and will use it.

In Doxygen, code bits such as MediaService::onCharacteristicChanged turn into hyperlinks to that item's documentation.  I had thought that putting them in backticks would disrupt that, but it doesn't, so I'm changing it throughout the documentation. 

I've added a note and a hyperlink to the MPRIS interface specification.

File: doc/02_architecture.md (hunk 15 above)
Comment (beroset) [reply doc/02_architecture.md:37]: Yes, each of the `@ref` entries refers to some other anchor in the documentation.  In some cases it refers to a glossary entry and in others, such as `@ref dbus` it refers to a page title.

File: doc/03_dbus.md (hunk 14 above)
Comment (beroset) [reply doc/03_dbus.md]: That's a good clarification.  I will change the footnote to say "Although StarfishOS supports D-Bus, it does not use the MPRIS interface for this purpose on that platform."

**Comment by beroset** [comment]
> Also I think it would be appropriate to link to https://asteroidos.org/wiki/ble-profiles/ when talking about BLE characteristics etc...

Good point!  Done.
 
> And one big thing I'd expect from this documentation is a clear distinction between "asteroidsyncserviced" the daemon and "asteroidsyncservice" the QML plugin which makes it easy to talk to "asteroidsyncserviced", I didn't find it particularly obvious from a first read.

I changed the last sentence of the first paragraph on the 02_architecture page to read "This project contains two major components: a QML plugin named `asteroidsyncserviceplugin` and a daemon with which it interacts named `asteroidsyncserviced`. " and also made that a little more explicit in the diagram.

**Comment by FlorentRevest** [comment]
Oops I only reviewed the incremental diff this time and it looks like I hit the merge button a bit too fast given that this contained commits about the service file...

**Pull request merged**